### PR TITLE
Improve Go SDK efficiency and concurrency

### PR DESCRIPTION
**Pull request by aludvik**
See commit message for high-level details.

The following diagram shows roughly how it works now:

image

### CODE REVIEW

File: sdk/go/src/sawtooth_sdk/messaging/connection.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	return nil
+}
+
+// SendNewMsg creates a new validator message, assigns a new correlation id,
+// serializes it, and sends it. It returns the correlation id created.
+func (self *Connection) SendNewMsg(t validator_pb2.Message_MessageType, c []byte) (corrId string, err error) {
+	return self.SendNewMsgTo("", t, c)
+}
+
+// SendNewMsgTo sends a new message to the i
```
Comment (vaporos) [inline sdk/go/src/sawtooth_sdk/messaging/connection.go]: to the i?

File: sdk/go/src/sawtooth_sdk/processor/handler.go [hunk 2]
```diff
@@ -15,8 +15,15 @@
  * ------------------------------------------------------------------------------
  */
 
-// Package processor provides a high-level generic transaction processor that
+// Package processor defines:
+// 1. A TransactionHandler interface to be used to create new transaction
+// families.
+//
+// 2. A high-level, general purpose, mutli-threaded TransactionProcessor that
```
Comment (vaporos) [inline sdk/go/src/sawtooth_sdk/processor/handler.go]: Typo - "mutli"

**Comment by aludvik** [comment]
Cleaned up comment-docs on Connection.go and fixed handler typo.

File: sdk/go/src/sawtooth_sdk/messaging/connection.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+		CorrelationId: corrId,
+		Content:       c,
+	}
+	return proto.Marshal(msg)
+}
+
+// LoadMsg deserializes a validator message
+func LoadMsg(data []byte) (msg *validator_pb2.Message, err error) {
+	msg = &validator_pb2.Message{}
+	err = proto.Unmarshal(data, msg)
```
Comment (peterschwarz) [inline sdk/go/src/sawtooth_sdk/messaging/connection.go]: Should something be done with this `err` value?

File: sdk/examples/intkey_go/src/sawtooth_intkey/main.go [hunk 4]
```diff
@@ -32,11 +32,15 @@ func main() {
...
 	logger := logging.Get()
-	logger.SetLevel(logging.INFO)
+	logger.SetLevel(logging.DEBUG)
 
 	prefix := intkey.Hexdigest("intkey")[:6]
 	handler := intkey.NewIntkeyHandler(prefix)
 	processor := processor.NewTransactionProcessor(endpoint)
 	processor.AddHandler(handler)
-	processor.Start()
+	processor.SetThreadCount(5)
```
Comment (peterschwarz) [inline sdk/examples/intkey_go/src/sawtooth_intkey/main.go]: 5 seems a bit arbitrary. Can this be tied to the number of cpu's?

File: sdk/go/src/sawtooth_sdk/processor/processor.go [hunk 5]
```diff
@@ -29,93 +30,266 @@ import (
 
 var logger *logging.Logger = logging.Get()
 
+const DEFAULT_THREAD_COUNT = 10
```
Comment (peterschwarz) [inline sdk/go/src/sawtooth_sdk/processor/processor.go]: Perhaps this default could be the number of CPU's (so probably not a `const`)

File: sdk/go/src/sawtooth_sdk/messaging/connection.go (hunk 3 above)
Comment (aludvik) [inline sdk/go/src/sawtooth_sdk/messaging/connection.go]: It is being returned using named return values: https://tour.golang.org/basics/7

**Comment by aludvik** [comment]
Set default number of threads equal to GOMAXPROCS, which defaults to the number of logical CPUs available.

Handled receiving messages out of order.

**Pull request merged**